Assets/Scripts/CharacterControllers/Melody/States/AttackState.cs

```diff
@@ -0,0 +1,28 @@
+﻿namespace Melody.States
+{
+    public class AttackState : MelodyState
+    {
+        public AttackState(MelodyController controller) : base(controller) { }
+
+        protected override void Enter()
+        {
+            melodyController.animator.SetTrigger("Attack");
+        }
+
+        public override void OnUpdate(float time)
+        {
+            base.OnUpdate(time);
+            if (melodyController.animator.IsInTransition(0) && melodyController.animator.GetCurrentAnimatorStateInfo(0).IsName("Attack"))
```

Comment:
We can avoid a string comparison here using [AnimatorState.nameHash](https://docs.unity3d.com/ScriptReference/Animations.AnimatorState-nameHash.html).
So you'd cache that int hash in Enter() and compare it with `sGetCurrentAnimatorStateInfo(0).hortNameHash`, I think.
Over optimization? Update is called a lot presumably.